File: src/page/MainLocale.java
package page;

import java.io.PrintStream;
import java.util.Locale;
import java.util.Map;
import java.util.Map.Entry;

import common.CommonStatic;
import common.CommonStatic.Lang;
import java.util.MissingResourceException;
import java.util.ResourceBundle;
import java.util.TreeMap;

import io.Writer;
import main.MainBCU;
import page.basis.ComboListTable;
import page.battle.BattleInfoPage;
import page.info.HeadTable;
import page.info.StageTable;
import page.info.edit.StageEditPage;
import page.info.filter.EnemyListTable;
import page.info.filter.UnitListTable;
// FIXME include page.pack import page.pack.EREditPage;
import utilpc.Interpret;

public strictfp class MainLocale {

	public static final Map<String, MainLocale> NAMP = new TreeMap<>();
	public static final Map<String, TTT> TMAP = new TreeMap<>();
	public static final String[] LOC_NAME = { "English", "\u4E2D\u6587", "\uD55C\uAD6D\uC5B4", "Japanese" };
	public static final String[] RENN = { "page", "info", "internet", "util" };
	private static final ResourceBundle[] RENS = new ResourceBundle[4];

	static {
		for (int i = 0; i < 4; i++)
			RENS[i] = ResourceBundle.getBundle((MainBCU.WRITE ? "src/" : "") + "page/" + RENN[i], Locale.ENGLISH);
	}

	public static boolean exLang, exTTT;

	public static TTT addTTT(String loc, String page, String text, String cont) {
		TTT ttt = TMAP.get(loc);
		if (ttt == null)
			TMAP.put(loc, ttt = new TTT());
		if (text.equals("*"))
			ttt.tttp.put(page, cont);
		else if (page.equals("*"))
			ttt.ttts.put(text, cont);
		else {
			if (!ttt.ttt.containsKey(page))
				ttt.ttt.put(page, new TreeMap<String, String>());
			ttt.ttt.get(page).put(text, cont);
		}
		return ttt;
	}

	public static String getLoc(int loc, String key) {
		if (loc >= 0 && loc < 4) {
			String loci = RENN[loc] + "_";
			String locl = loci + langCode();
			if (NAMP.containsKey(locl) && NAMP.get(locl).contains(key)) {
				String str = NAMP.get(loci + langCode()).get(key);
				if (str.equals("(null)"))
					str = RENS[loc].getString(key);
				String[] strs = str.split("#");
				if (strs.length == 1)
					return str;
				for (int i = 1; i < strs.length; i += 2)
					strs[i] = getLoc(loc, strs[i]);
				String ans = "";
				for (int i = 0; i < strs.length; i++)
					ans += strs[i];
				return ans;
			}
			try {
				return exLang ? "[" + loci + key + "]" : RENS[loc].getString(key);
			} catch (MissingResourceException e) {
				return key;
			}
		}
		return key;
	}

	public static String[] getLoc(int loc, String... strs) {
		String[] ans = new String[strs.length];
		for (int i = 0; i < ans.length; i++)
			ans[i] = getLoc(loc, strs[i]);
		return ans;
	}

	public static String[] getLoc(int loc, String pre, int max) {
		String[] ans = new String[max];
		for (int i = 0; i < ans.length; i++)
			ans[i] = getLoc(loc, pre + i);
		return ans;
	}

	public static void saveWorks() {
		for (String loc : Lang.LOC_CODE) {
			TTT ttt = TMAP.get(loc);
			if (ttt != null && ttt.edited)
				ttt.write(Writer.newFile("./lib/lang/" + loc + "/tutorial.txt"));
			for (int i = 0; i < 4; i++) {
				MainLocale ml = NAMP.get(RENN[i] + "_" + loc);
				if (ml != null && ml.edited)
					ml.write(Writer.newFile("./lib/lang/" + loc + "/" + RENN[i] + ".properties"));
			}
		}
	}

	protected static String getTTT(String page, String text) {
		String loc = langCode();
		String ans = null;
		if (TMAP.containsKey(loc))
			ans = TMAP.get(loc).getTTT(page, text);
		if (ans != null)
			return ans;
		if (exTTT)
			return "[" + page + "_" + text + "]";
		loc = Lang.LOC_CODE[0];
		if (TMAP.containsKey(loc))
			return TMAP.get(loc).getTTT(page, text);
		return null;
	}

	protected static void redefine() {
		Interpret.redefine();
		EnemyListTable.redefine();
		UnitListTable.redefine();
		ComboListTable.redefine();
		StageTable.redefine();
		HeadTable.redefine();
		BattleInfoPage.redefine();
		StageEditPage.redefine();
		// FIXME inlucde page.pack EREditPage.redefine();
	}

	protected static void setLoc(int i, String key, String value) {
		if (i < 0)
			return;
		String loc = RENN[i] + "_" + langCode();
		MainLocale ml = NAMP.get(loc);
		if (ml == null)
			NAMP.put(loc, ml = new MainLocale(loc));
		ml.res.put(key, value);
		ml.edited = true;
	}

	protected static void setTTT(String loc, String info, String str) {
		addTTT(langCode(), loc, info, str).edited = true;
	}

	private static String langCode() {
		return CommonStatic.Lang.LOC_CODE[CommonStatic.getConfig().lang];
	}

	public final Map<String, String> res = new TreeMap<>();

	public boolean edited;

	public MainLocale(String str) {
		NAMP.put(str, this);
	}

	public void write(PrintStream ps) {
		for (Entry<String, String> ent : res.entrySet())
			ps.println(ent.getKey() + "\t" + ent.getValue());
	}

	private boolean contains(String str) {
		return res.containsKey(str);
	}

	private String get(String str) {
		if (res.containsKey(str))
			return res.get(str);
		return "!" + str + "!";
	}

}

class TTT {

	public final Map<String, String> tttp = new TreeMap<>();
	public final Map<String, String> ttts = new TreeMap<>();
	public final Map<String, Map<String, String>> ttt = new TreeMap<>();

	public boolean edited;

	public void write(PrintStream ps) {
		if (ps == null)
			return;
		for (Entry<String, String> ent : tttp.entrySet())
			ps.println(ent.getKey() + "\t*\t" + ent.getValue());
		for (Entry<String, String> ent : ttts.entrySet())
			ps.println("*\t" + ent.getKey() + "\t" + ent.getValue());
		for (Entry<String, Map<String, String>> ent : ttt.entrySet())
			for (Entry<String, String> snt : ent.getValue().entrySet())
				ps.println(ent.getKey() + "\t" + snt.getKey() + "\t" + snt.getValue());
	}

	protected String getTTT(String page, String text) {
		if (ttt.get(page) != null) {
			String strs = ttt.get(page).get(text);
			if (strs != null)
				return strs;
		}
		String strt = ttts.get(text);
		if (strt != null)
			return strt;
		String strp = tttp.get(page);
		if (strp != null)
			return strp;
		return null;
	}
}

File: src/page/MainLocale.kt
package page

import common.CommonStatic
import common.io.assets.Admin
import common.io.assets.Admin.StaticPermitted
import common.io.assets.AssetLoader
import common.io.assets.AssetLoader.AssetHeader
import common.io.assets.AssetLoader.AssetHeader.AssetEntry
import common.io.json.JsonEncoder
import common.io.json.Test
import common.io.json.Test.JsonTest_0.JsonD
import common.io.json.Test.JsonTest_2
import common.pack.Source.AnimLoader
import common.pack.Source.ResourceLocation
import common.pack.Source.SourceAnimLoader
import common.pack.Source.SourceAnimSaver
import common.pack.Source.Workspace
import common.pack.Source.ZipSource
import common.util.stage.EStage
import common.util.stage.StageMap
import common.util.stage.StageMap.StageMapInfo
import common.util.unit.UnitLevel
import io.BCPlayer
import io.BCUWriter
import main.MainBCU
import page.JL
import page.anim.AnimBox
import page.basis.ComboListTable
import page.battle.BattleInfoPage
import page.info.HeadTable
import page.info.StageTable
import page.info.edit.StageEditPage
import page.info.filter.EnemyListTable
import page.info.filter.UnitListTable
import page.support.ListJtfPolicy
import page.support.SortTable
import page.view.ViewBox
import page.view.ViewBox.Conf
import page.view.ViewBox.Controller
import page.view.ViewBox.VBExporter
import utilpc.Interpret
import java.io.PrintStream
import java.util.*

// FIXME include page.pack import page.pack.EREditPage;
@Strictfp
class MainLocale(str: String) {
    companion object {
        val NAMP: MutableMap<String, MainLocale> = TreeMap<String, MainLocale>()
        val TMAP: MutableMap<String, TTT> = TreeMap<String, TTT>()
        val LOC_NAME = arrayOf("English", "\u4E2D\u6587", "\uD55C\uAD6D\uC5B4", "Japanese")
        val RENN = arrayOf("page", "info", "internet", "util")
        private val RENS: Array<ResourceBundle?> = arrayOfNulls<ResourceBundle>(4)
        var exLang = false
        var exTTT = false
        fun addTTT(loc: String, page: String, text: String, cont: String): TTT? {
            var ttt = TMAP[loc]
            if (ttt == null) TMAP[loc] = TTT().also { ttt = it }
            if (text == "*") ttt!!.tttp[page] = cont else if (page == "*") ttt!!.ttts[text] = cont else {
                if (!ttt!!.ttt.containsKey(page)) ttt!!.ttt[page] = TreeMap<String, String>()
                ttt!!.ttt[page]!![text] = cont
            }
            return ttt
        }

        fun getLoc(loc: Int, key: String?): String? {
            if (loc >= 0 && loc < 4) {
                val loci = RENN[loc] + "_"
                val locl = loci + langCode()
                if (NAMP.containsKey(locl) && NAMP[locl]!!.contains(key)) {
                    var str = NAMP[loci + langCode()]!![key]
                    if (str == "(null)") str = RENS[loc].getString(key)
                    val strs: Array<String?> = str!!.split("#").toTypedArray()
                    if (strs.size == 1) return str
                    run {
                        var i = 1
                        while (i < strs.size) {
                            strs[i] = getLoc(loc, strs[i])
                            i += 2
                        }
                    }
                    var ans: String? = ""
                    for (i in strs.indices) ans += strs[i]
                    return ans
                }
                return try {
                    if (exLang) "[$loci$key]" else RENS[loc].getString(key)
                } catch (e: MissingResourceException) {
                    key
                }
            }
            return key
        }

        fun getLoc(loc: Int, vararg strs: String?): Array<String?> {
            val ans = arrayOfNulls<String>(strs.size)
            for (i in ans.indices) ans[i] = getLoc(loc, strs[i])
            return ans
        }

        fun getLoc(loc: Int, pre: String, max: Int): Array<String?> {
            val ans = arrayOfNulls<String>(max)
            for (i in ans.indices) ans[i] = getLoc(loc, pre + i)
            return ans
        }

        fun saveWorks() {
            for (loc in CommonStatic.Lang.Companion.LOC_CODE) {
                val ttt = TMAP[loc]
                if (ttt != null && ttt.edited) ttt.write(BCUWriter.newFile("./lib/lang/$loc/tutorial.txt"))
                for (i in 0..3) {
                    val ml = NAMP[RENN[i] + "_" + loc]
                    if (ml != null && ml.edited) ml.write(BCUWriter.newFile("./lib/lang/" + loc + "/" + RENN[i] + ".properties"))
                }
            }
        }

        fun getTTT(page: String, text: String): String? {
            var loc = langCode()
            var ans: String? = null
            if (TMAP.containsKey(loc)) ans = TMAP[loc]!!.getTTT(page, text)
            if (ans != null) return ans
            if (exTTT) return "[" + page + "_" + text + "]"
            loc = CommonStatic.Lang.Companion.LOC_CODE.get(0)
            return if (TMAP.containsKey(loc)) TMAP[loc]!!.getTTT(page, text) else null
        }

        fun redefine() {
            Interpret.redefine()
            EnemyListTable.Companion.redefine()
            UnitListTable.Companion.redefine()
            ComboListTable.Companion.redefine()
            StageTable.Companion.redefine()
            HeadTable.Companion.redefine()
            BattleInfoPage.Companion.redefine()
            StageEditPage.Companion.redefine()
            // FIXME inlucde page.pack EREditPage.redefine();
        }

        fun setLoc(i: Int, key: String?, value: String) {
            if (i < 0) return
            val loc = RENN[i] + "_" + langCode()
            var ml = NAMP[loc]
            if (ml == null) NAMP[loc] = MainLocale(loc).also { ml = it }
            ml!!.res[key] = value
            ml!!.edited = true
        }

        fun setTTT(loc: String, info: String, str: String) {
            addTTT(langCode(), loc, info, str)!!.edited = true
        }

        private fun langCode(): String {
            return CommonStatic.Lang.Companion.LOC_CODE.get(CommonStatic.getConfig().lang)
        }

        init {
            for (i in 0..3) RENS[i] = ResourceBundle.getBundle((if (MainBCU.WRITE) "src/" else "") + "page/" + RENN[i], Locale.ENGLISH)
        }
    }

    val res: MutableMap<String?, String> = TreeMap<String, String>()
    var edited = false
    fun write(ps: PrintStream) {
        for ((key, value) in res) ps.println(key.toString() + "\t" + value)
    }

    private operator fun contains(str: String?): Boolean {
        return res.containsKey(str)
    }

    private operator fun get(str: String?): String? {
        return if (res.containsKey(str)) res[str] else "!$str!"
    }

    init {
        NAMP[str] = this
    }
}

class TTT {
    val tttp: MutableMap<String, String> = TreeMap<String, String>()
    val ttts: MutableMap<String, String> = TreeMap<String, String>()
    val ttt: MutableMap<String, MutableMap<String, String>?> = TreeMap<String, Map<String, String>>()
    var edited = false
    fun write(ps: PrintStream?) {
        if (ps == null) return
        for ((key, value) in tttp) ps.println("$key\t*\t$value")
        for ((key, value) in ttts) ps.println("*\t$key\t$value")
        for ((key, value) in ttt) for ((key1, value1) in value!!) ps.println(key + "\t" + key1 + "\t" + value1)
    }

    fun getTTT(page: String, text: String): String? {
        if (ttt[page] != null) {
            val strs = ttt[page]!![text]
            if (strs != null) return strs
        }
        val strt = ttts[text]
        return strt ?: tttp[page]
    }
}
